Graph: Nonnull tooltip hover mode

kvc0 opened the pull request:
## Add a Nonnull tooltip hover mode

This adds a Nonnull option to hover tooltip modes.
Nonnull hides series in a tooltip when they would be displayed as null.
This is useful when you have sparsely reported series and only care which values were reported at a given time column.

All tests pass.


**What this PR does / why we need it**:
<img width="1227" alt="image" src="https://user-images.githubusercontent.com/3454741/100298024-e1f2cf00-2f44-11eb-99e4-33de6ebfdd52.png">

This new option hides series with null value at the currently hovered time column from the data tooltip in graph panels.

When a time series has sparse data - e.g., reporting infrequent errors by `device_id`, and potentially _many_ distinct series across a dashboard's time range - e.g., a `device_id`, a tooltip becomes quickly unreadable if you don't sort by value descending, or unresponsive if you sort 10,000 mostly-null series by value descending for every pixel the mouse moves.

This gives dashboard authors the ability to choose whether null valued series show up in a hover tooltip.


**Which issue(s) this PR fixes**:
There are many "closed" issues from people who appear to be asking for this but they are either misunderstood or fall into the trap of also mentioning the existing feature of "hide a series from the tooltip when ALL of its values happen to be 0 or null" which is different from this PR which focuses on sparse data.

Relates to:
https://github.com/grafana/grafana/issues/3094
https://github.com/grafana/grafana/issues/1381
https://github.com/grafana/grafana/issues/4623

This PR does not include a "Nonzero" option and it does not affect what is in the legend, so it does not seem to satisfy every aspect of each of these (and myriad other loosely related) issues.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=29405) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=29405) before we can accept your contribution.<br/><hr/>**Kenny** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=29405) it.</sub>

kvc0 committed:
Add nonnull hover mode

This adds a Nonnull option to hover tooltip modes.

Nonnull hides series in a tooltip when they would be displayed as null.

This is useful when you have sparsely reported series and only care which values were reported at a given time column.

oddlittlebird commented:
Doc change looks fine to me.

A dev will review the code when time allows.

WarriorOfWire commented:
Per stalebot, I’m pinging for review with a status update:  Ready for review.

WarriorOfWire commented:
This PR from November remains current and awaiting review.  Does stalebot need its thresholds updated to meet the expectation of review latency in this repository?

torkelo commented:
Closing this as we do not want to modify the old graph panel any more, 

Here is the new tooltip for the new time series panel:
https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx

anacrolix commented:
The new time series panel doesn't seem to have any tooltip options for the proposed mode.

WarriorOfWire commented:
@anacrolix yup. Grafana has ensured that I won’t port this feature to the new panel, but I’d be happy if someone else were to conjure the arcane incantations necessary to merge a minor feature pr into it!
